File: aeonmind/python/aeonmind/core/adaptive.py
"""
AeonMind Adaptive Meta-Learning — Python Implementation.

Implements L-BFGS (Limited-memory Broyden–Fletcher–Goldfarb–Shanno)
two-loop recursion for Hessian approximation with momentum, gradient
clipping, and adaptive learning rate scheduling.
"""

from __future__ import annotations

import math  # noqa: F401
from dataclasses import dataclass, field
from typing import Callable, List, Optional  # noqa: UP035

import numpy as np


@dataclass
class AdaptiveConfig:
    """Configuration for the AdaptiveMetaLearner."""
    learning_rate: float = 0.01
    memory_size: int = 10
    max_iterations: int = 1000
    tolerance: float = 1e-8
    gradient_clip: float = 1.0
    momentum: float = 0.9
    beta1: float = 0.9
    beta2: float = 0.999
    epsilon: float = 1e-8
    decay_rate: float = 0.95
    warmup_steps: int = 100
    use_lbfgs: bool = True


@dataclass
class LbfgsEntry:
    """Storage entry for L-BFGS two-loop recursion."""
    s: np.ndarray  # parameter difference
    y: np.ndarray  # gradient difference
    rho: float     # 1 / (y^T s)


@dataclass
class AdaptiveStep:
    """Result of a single optimization step."""
    iteration: int
    loss: float
    gradient_norm: float
    learning_rate: float
    step_size: float


@dataclass
class AdaptiveSummary:
    """Summary of the adaptive optimization process."""
    total_steps: int
    final_loss: float
    final_gradient_norm: float
    initial_loss: float
    best_loss: float
    best_iteration: int
    converged: bool
    learning_rate_history: List[float] = field(default_factory=list)  # noqa: UP006
    loss_history: List[float] = field(default_factory=list)  # noqa: UP006


class AdaptiveMetaLearner:
    """Adaptive Meta-Learner with L-BFGS two-loop recursion.

    Uses limited-memory BFGS for Hessian approximation, combined
    with momentum-based gradient descent and adaptive learning rate
    scheduling for efficient optimization.
    """

    def __init__(self, n_params: int, config: Optional[AdaptiveConfig] = None):  # noqa: UP045
        self.n_params = n_params
        self.config = config or AdaptiveConfig()
        self.parameters = np.random.randn(n_params) * 0.01
        self._velocity = np.zeros(n_params)
        self._moment_estimates_m = np.zeros(n_params)
        self._moment_estimates_v = np.zeros(n_params)
        self._lbfgs_history: List[LbfgsEntry] = []  # noqa: UP006
        self._prev_gradient: Optional[np.ndarray] = None  # noqa: UP045
        self._step_count = 0
        self._best_loss = float("inf")
        self._best_params: Optional[np.ndarray] = None  # noqa: UP045
        self._best_iteration = 0
        self._initial_loss: Optional[float] = None  # noqa: UP045
        self._loss_history: List[float] = []  # noqa: UP006
        self._lr_history: List[float] = []  # noqa: UP006

    @classmethod
    def with_parameters(cls, parameters: np.ndarray, config: Optional[AdaptiveConfig] = None) -> AdaptiveMetaLearner:  # noqa: UP045
        """Create a learner initialized with specific parameters."""
        learner = cls(len(parameters), config)
        learner.parameters = parameters.copy()
        return learner

    def lbfgs_direction(self, gradient: np.ndarray) -> np.ndarray:
        """Compute search direction using L-BFGS two-loop recursion.

        The two-loop recursion approximates the inverse Hessian-vector
        product without storing the full Hessian matrix.
        """
        direction = -gradient.copy()
        q = direction.copy()

        # First loop: traverse from most recent to oldest
        alphas = []
        for entry in reversed(self._lbfgs_history):
            alpha = entry.rho * np.dot(entry.s, q)
            alphas.append(alpha)
            q = q - alpha * entry.y

        # Apply initial Hessian approximation (scaled identity)
        if self._lbfgs_history:
            latest = self._lbfgs_history[-1]
            gamma = np.dot(latest.s, latest.y) / (np.dot(latest.y, latest.y) + 1e-10)
            q = gamma * q

        # Second loop: traverse from oldest to most recent
        for entry, alpha in zip(self._lbfgs_history, reversed(alphas)):
            beta = entry.rho * np.dot(entry.y, q)
            q = q + entry.s * (alpha - beta)

        return q

    def _update_lbfgs_history(self, s: np.ndarray, y: np.ndarray) -> None:
        """Update L-BFGS history with new curvature pair."""
        ys = np.dot(y, s)
        if ys > 1e-10:
            rho = 1.0 / ys
            self._lbfgs_history.append(LbfgsEntry(s=s, y=y, rho=rho))
            if len(self._lbfgs_history) > self.config.memory_size:
                self._lbfgs_history.pop(0)

    def step(self, gradient: np.ndarray) -> AdaptiveStep:
        """Perform a single optimization step.

        Combines L-BFGS direction with momentum and adaptive learning rate.
        """
        # Clip gradient
        grad_norm = np.linalg.norm(gradient)
        if grad_norm > self.config.gradient_clip:
            gradient = gradient * (self.config.gradient_clip / grad_norm)

        # Compute direction
        if self.config.use_lbfgs and len(self._lbfgs_history) > 0:
            direction = self.lbfgs_direction(gradient)
        else:
            direction = -gradient

        # Update momentum estimates (Adam-style)
        self._moment_estimates_m = (
            self.config.beta1 * self._moment_estimates_m
            + (1 - self.config.beta1) * direction
        )
        self._moment_estimates_v = (
            self.config.beta2 * self._moment_estimates_v
            + (1 - self.config.beta2) * direction ** 2
        )

        # Bias correction
        m_hat = self._moment_estimates_m / (1 - self.config.beta1 ** (self._step_count + 1))
        v_hat = self._moment_estimates_v / (1 - self.config.beta2 ** (self._step_count + 1))

        # Compute update
        lr = self._current_learning_rate()
        update = lr * m_hat / (np.sqrt(v_hat) + self.config.epsilon)

        # Update L-BFGS history
        if self._prev_gradient is not None:
            s = -update  # parameter change
            y = gradient - self._prev_gradient  # gradient change
            self._update_lbfgs_history(s, y)

        # Apply velocity with momentum
        self._velocity = self.config.momentum * self._velocity + update
        self.parameters = self.parameters - self._velocity

        self._prev_gradient = gradient.copy()
        self._step_count += 1

        return AdaptiveStep(
            iteration=self._step_count,
            loss=0.0,  # caller should compute
            gradient_norm=float(np.linalg.norm(gradient)),
            learning_rate=lr,
            step_size=float(np.linalg.norm(update)),
        )

    def optimize(
        self,
        loss_fn: Callable[[np.ndarray], float],
        grad_fn: Callable[[np.ndarray], np.ndarray],
        callback: Optional[Callable[[int, float, float], None]] = None,  # noqa: UP045
    ) -> AdaptiveSummary:
        """Run full optimization loop."""
        self._initial_loss = loss_fn(self.parameters)
        self._best_loss = self._initial_loss

        for i in range(self.config.max_iterations):
            loss = loss_fn(self.parameters)
            gradient = grad_fn(self.parameters)

            step_result = self.step(gradient)
            step_result.loss = loss

            self._loss_history.append(loss)
            self._lr_history.append(step_result.learning_rate)

            if loss < self._best_loss:
                self._best_loss = loss
                self._best_params = self.parameters.copy()
                self._best_iteration = i

            if callback:
                callback(i, loss, step_result.gradient_norm)

            if step_result.gradient_norm < self.config.tolerance:
                break

        return AdaptiveSummary(
            total_steps=self._step_count,
            final_loss=self._loss_history[-1] if self._loss_history else 0.0,
            final_gradient_norm=step_result.gradient_norm if 'step_result' in dir() else 0.0,
            initial_loss=self._initial_loss or 0.0,
            best_loss=self._best_loss,
            best_iteration=self._best_iteration,
            converged=step_result.gradient_norm < self.config.tolerance if 'step_result' in dir() else False,  # noqa: E501
            learning_rate_history=self._lr_history.copy(),
            loss_history=self._loss_history.copy(),
        )

    def _current_learning_rate(self) -> float:
        """Compute current learning rate with warmup and decay."""
        if self._step_count < self.config.warmup_steps:
            scale = (self._step_count + 1) / self.config.warmup_steps
            return self.config.learning_rate * scale
        return self.config.learning_rate * (self.config.decay_rate ** (self._step_count - self.config.warmup_steps))  # noqa: E501

    def adapt_learning_rate(self, loss: float) -> float:
        """Adapt learning rate based on loss progress."""
        if len(self._loss_history) >= 2:
            recent = self._loss_history[-5:] if len(self._loss_history) >= 5 else self._loss_history
            if len(recent) >= 2 and recent[-1] > recent[-2]:
                return self.config.learning_rate * 0.5
        return self.config.learning_rate

    def parameters_array(self) -> np.ndarray:
        """Return a copy of current parameters."""
        return self.parameters.copy()

    def summary(self) -> AdaptiveSummary:
        """Get summary of the optimization process."""
        return AdaptiveSummary(
            total_steps=self._step_count,
            final_loss=self._loss_history[-1] if self._loss_history else 0.0,
            final_gradient_norm=float(np.linalg.norm(self._prev_gradient)) if self._prev_gradient is not None else 0.0,  # noqa: E501
            initial_loss=self._initial_loss or 0.0,
            best_loss=self._best_loss,
            best_iteration=self._best_iteration,
            converged=False,
            learning_rate_history=self._lr_history.copy(),
            loss_history=self._loss_history.copy(),
        )

    def reset(self) -> None:
        """Reset the learner state."""
        self.parameters = np.random.randn(self.n_params) * 0.01
        self._velocity = np.zeros(self.n_params)
        self._moment_estimates_m = np.zeros(self.n_params)
        self._moment_estimates_v = np.zeros(self.n_params)
        self._lbfgs_history = []
        self._prev_gradient = None
        self._step_count = 0
        self._best_loss = float("inf")
        self._best_params = None
        self._best_iteration = 0
        self._initial_loss = None
        self._loss_history = []
        self._lr_history = []

File: aeonmind/rust/src/adaptive.rs
/*
AeonMind Adaptive Meta-Learning — L-BFGS Optimization.

Implements adaptive hyperparameter tuning with:
    - L-BFGS two-loop recursion for Hessian approximation
    - Adaptive learning rate with momentum
    - Gradient clipping
    - Weight decay regularization
    - History-based curvature estimation

Part of the Tranc3 Infinity Ecosystem.
*/

use serde::{Deserialize, Serialize};

// ─── Configuration ──────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdaptiveConfig {
    pub n_parameters: usize,
    pub learning_rate: f64,
    pub history_size: usize,
    pub bounds: (f64, f64),
    pub adaptive_lr: bool,
    pub momentum: f64,
    pub weight_decay: f64,
    pub gradient_clip: f64,
}

impl Default for AdaptiveConfig {
    fn default() -> Self {
        Self {
            n_parameters: 10,
            learning_rate: 0.01,
            history_size: 10,
            bounds: (-10.0, 10.0),
            adaptive_lr: true,
            momentum: 0.9,
            weight_decay: 0.0001,
            gradient_clip: 5.0,
        }
    }
}

// ─── L-BFGS History Entry ───────────────────────────────────

#[derive(Debug, Clone)]
struct LbfgsEntry {
    s: Vec<f64>, // Parameter difference
    y: Vec<f64>, // Gradient difference
    rho: f64,    // 1 / (y^T s)
}

// ─── Adaptive Step ──────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdaptiveStep {
    pub iteration: usize,
    pub cost: f64,
    pub gradient_norm: f64,
    pub learning_rate: f64,
    pub step_size: f64,
}

// ─── Adaptive Summary ───────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdaptiveSummary {
    pub n_parameters: usize,
    pub learning_rate: f64,
    pub history_size: usize,
    pub iterations: usize,
    pub final_cost: Option<f64>,
}

// ─── Adaptive Meta-Learner ──────────────────────────────────

pub struct AdaptiveMetaLearner {
    config: AdaptiveConfig,
    parameters: Vec<f64>,
    velocity: Vec<f64>,
    history: Vec<LbfgsEntry>,
    prev_gradient: Option<Vec<f64>>,
    iteration: usize,
    current_cost: Option<f64>,
    step_history: Vec<AdaptiveStep>,
}

impl AdaptiveMetaLearner {
    pub fn new(config: AdaptiveConfig) -> Self {
        let mut rng = rand::thread_rng();
        let parameters: Vec<f64> = (0..config.n_parameters)
            .map(|_| {
                let (lo, hi) = config.bounds;
                rand::Rng::gen_range(&mut rng, lo..hi)
            })
            .collect();

        Self {
            velocity: vec![0.0; config.n_parameters],
            history: Vec::new(),
            prev_gradient: None,
            iteration: 0,
            current_cost: None,
            step_history: Vec::new(),
            config,
            parameters,
        }
    }

    pub fn with_parameters(config: AdaptiveConfig, parameters: Vec<f64>) -> Self {
        let n = parameters.len();
        let mut learner = Self::new(AdaptiveConfig {
            n_parameters: n,
            ..config
        });
        learner.parameters = parameters;
        learner
    }

    /// Compute L-BFGS search direction using two-loop recursion.
    pub fn lbfgs_direction(&self, gradient: &[f64]) -> Vec<f64> {
        let mut q = gradient.to_vec();

        // First loop (backwards through history)
        let m = self.history.len();
        let mut alphas = vec![0.0; m];

        for i in (0..m).rev() {
            alphas[i] = self.history[i].rho * dot(&self.history[i].s, &q);
            for (qj, yj) in q.iter_mut().zip(self.history[i].y.iter()) {
                *qj -= alphas[i] * yj;
            }
        }

        // Initial Hessian approximation (scaled identity)
        let gamma = if m > 0 {
            let last = &self.history[m - 1];
            let y_norm_sq = dot(&last.y, &last.y);
            if y_norm_sq > 1e-10 {
                dot(&last.s, &last.y) / y_norm_sq
            } else {
                1.0
            }
        } else {
            1.0
        };

        let mut r: Vec<f64> = q.iter().map(|&v| gamma * v).collect();

        // Second loop (forwards through history)
        for (alpha_i, hist_item) in alphas.iter().zip(self.history.iter()) {
            let beta = hist_item.rho * dot(&hist_item.y, &r);
            for (rj, sj) in r.iter_mut().zip(hist_item.s.iter()) {
                *rj += sj * (alpha_i - beta);
            }
        }

        r
    }

    /// Perform one optimization step.
    pub fn step(&mut self, gradient: &[f64]) -> AdaptiveStep {
        let grad_norm = norm(gradient);

        // Gradient clipping
        let clipped: Vec<f64> = if grad_norm > self.config.gradient_clip {
            gradient
                .iter()
                .map(|&g| g * self.config.gradient_clip / grad_norm)
                .collect()
        } else {
            gradient.to_vec()
        };

        // Compute L-BFGS direction
        let direction = self.lbfgs_direction(&clipped);

        // Snapshot parameters before update so we can compute the true s vector
        let prev_params = self.parameters.clone();

        // Momentum update
        let lr = self.config.learning_rate;
        for (vel, dir) in self.velocity.iter_mut().zip(direction.iter()) {
            *vel = self.config.momentum * *vel - lr * dir;
        }

        // Parameter update with weight decay
        let (lo, hi) = self.config.bounds;
        for (param, vel) in self.parameters.iter_mut().zip(self.velocity.iter()) {
            *param += vel - self.config.weight_decay * *param;
            // Apply bounds
            *param = param.max(lo).min(hi);
        }

        // Update L-BFGS history with true parameter difference x_{k+1} - x_k
        if let Some(ref prev_grad) = self.prev_gradient {
            let s: Vec<f64> = self
                .parameters
                .iter()
                .zip(prev_params.iter())
                .map(|(p_new, p_old)| p_new - p_old)
                .collect();
            let y: Vec<f64> = clipped
                .iter()
                .zip(prev_grad.iter())
                .map(|(&g, &pg)| g - pg)
                .collect();
            let ys = dot(&y, &s);
            if ys > 1e-10 {
                self.history.push(LbfgsEntry {
                    s,
                    y,
                    rho: 1.0 / ys,
                });
                if self.history.len() > self.config.history_size {
                    self.history.remove(0);
                }
            }
        }

        self.prev_gradient = Some(clipped);
        self.iteration += 1;

        // Adaptive learning rate
        if self.config.adaptive_lr {
            self.adapt_learning_rate();
        }

        let step_size = norm(&self.velocity);

        let step = AdaptiveStep {
            iteration: self.iteration,
            cost: self.current_cost.unwrap_or(0.0),
            gradient_norm: grad_norm,
            learning_rate: self.config.learning_rate,
            step_size,
        };
        self.step_history.push(step.clone());
        step
    }

    /// Run full optimization loop.
    pub fn optimize<F, G>(
        &mut self,
        cost_fn: F,
        gradient_fn: G,
        n_iterations: usize,
    ) -> Vec<AdaptiveStep>
    where
        F: Fn(&[f64]) -> f64 + Copy,
        G: Fn(&[f64]) -> Vec<f64> + Copy,
    {
        let mut steps = Vec::new();
        for _ in 0..n_iterations {
            let cost = cost_fn(&self.parameters);
            self.current_cost = Some(cost);
            let gradient = gradient_fn(&self.parameters);
            let step = self.step(&gradient);
            steps.push(step);
        }
        steps
    }

    /// Adapt learning rate based on progress.
    pub fn adapt_learning_rate(&mut self) {
        if self.step_history.len() >= 2 {
            let current = &self.step_history[self.step_history.len() - 1];
            let prev = &self.step_history[self.step_history.len() - 2];

            if current.gradient_norm > prev.gradient_norm * 1.2 {
                self.config.learning_rate *= 0.8;
            } else if current.gradient_norm < prev.gradient_norm * 0.8 {
                self.config.learning_rate = (self.config.learning_rate * 1.05).min(0.1);
            }
            self.config.learning_rate = self.config.learning_rate.max(1e-6);
        }
    }

    /// Get current parameters.
    pub fn parameters_array(&self) -> &[f64] {
        &self.parameters
    }

    /// Get summary.
    pub fn summary(&self) -> AdaptiveSummary {
        AdaptiveSummary {
            n_parameters: self.config.n_parameters,
            learning_rate: self.config.learning_rate,
            history_size: self.config.history_size,
            iterations: self.iteration,
            final_cost: self.current_cost,
        }
    }
}

// ─── Helper Functions ───────────────────────────────────────

fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b.iter()).map(|(&x, &y)| x * y).sum()
}

fn norm(v: &[f64]) -> f64 {
    v.iter().map(|x| x * x).sum::<f64>().sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_learner_creation() {
        let config = AdaptiveConfig::default();
        let learner = AdaptiveMetaLearner::new(config);
        assert_eq!(learner.parameters.len(), 10);
    }

    #[test]
    fn test_step() {
        let config = AdaptiveConfig {
            n_parameters: 5,
            ..Default::default()
        };
        let mut learner = AdaptiveMetaLearner::new(config);
        let gradient = vec![0.1; 5];
        let step = learner.step(&gradient);
        assert_eq!(step.iteration, 1);
    }

    #[test]
    fn test_lbfgs_direction() {
        let config = AdaptiveConfig {
            n_parameters: 5,
            ..Default::default()
        };
        let learner = AdaptiveMetaLearner::new(config);
        let gradient = vec![0.5; 5];
        let direction = learner.lbfgs_direction(&gradient);
        assert_eq!(direction.len(), 5);
    }

    #[test]
    fn test_gradient_clipping() {
        let config = AdaptiveConfig {
            n_parameters: 3,
            gradient_clip: 1.0,
            ..Default::default()
        };
        let mut learner = AdaptiveMetaLearner::new(config);
        let gradient = vec![100.0; 3];
        let step = learner.step(&gradient);
        // gradient_norm reports the pre-clip magnitude for convergence monitoring.
        // Input [100, 100, 100] has L2 norm = 100*sqrt(3) ≈ 173.2, well above clip=1.0.
        let expected_norm = (100.0_f64.powi(2) * 3.0).sqrt();
        assert!((step.gradient_norm - expected_norm).abs() < 1e-6);
    }

    #[test]
    fn test_momentum() {
        let config = AdaptiveConfig {
            n_parameters: 3,
            momentum: 0.9,
            ..Default::default()
        };
        let mut learner = AdaptiveMetaLearner::new(config);
        let grad1 = vec![1.0; 3];
        learner.step(&grad1);
        let grad2 = vec![1.0; 3];
        learner.step(&grad2);
        // With momentum, velocity should accumulate
        assert!(learner.velocity.iter().any(|&v| v != 0.0));
    }

    #[test]
    fn test_with_parameters() {
        let config = AdaptiveConfig::default();
        let params = vec![1.0, 2.0, 3.0];
        let learner = AdaptiveMetaLearner::with_parameters(config, params.clone());
        assert_eq!(learner.parameters, params);
    }

    #[test]
    fn test_quadratic_optimization() {
        let config = AdaptiveConfig {
            n_parameters: 3,
            learning_rate: 0.1,
            momentum: 0.0,
            adaptive_lr: false,
            ..Default::default()
        };
        // Use a deterministic starting point to avoid flakiness from random init.
        let mut learner = AdaptiveMetaLearner::with_parameters(config, vec![3.0, -3.0, 3.0]);
        // Minimize f(x) = sum(x_i^2)
        for _ in 0..100 {
            let grad: Vec<f64> = learner.parameters.iter().map(|&x| 2.0 * x).collect();
            learner.step(&grad);
        }
        // Parameters should converge near zero.
        let param_norm: f64 = learner.parameters.iter().map(|x| x * x).sum::<f64>().sqrt();
        assert!(param_norm < 1.0);
    }

    #[test]
    fn test_summary() {
        let config = AdaptiveConfig::default();
        let learner = AdaptiveMetaLearner::new(config);
        let summary = learner.summary();
        assert_eq!(summary.n_parameters, 10);
    }
}
